rust/log-service/src/lib.rs

```diff
@@ -967,12 +792,228 @@ impl LogServer {
                     Status::new(err.code().into(), format!("Failed to init cursor: {}", err))
                 })?;
         }
+        let mut need_to_compact = self.need_to_compact.lock();
+        if let Entry::Occupied(mut entry) = need_to_compact.entry(collection_id) {
+            let rollup = entry.get_mut();
+            rollup.start_log_position = std::cmp::max(
+                rollup.start_log_position,
+                LogPosition::from_offset(request.log_offset as u64),
+            );
+            if rollup.start_log_position >= rollup.limit_log_position {
+                entry.remove();
+            }
+        }
         Ok(Response::new(UpdateCollectionLogOffsetResponse {}))
     }
-}
 
-#[async_trait::async_trait]
-impl LogService for LogServer {
+    #[tracing::instrument(skip(self), err(Display))]
+    async fn cached_get_all_collection_info_to_compact(
+        &self,
+        request: GetAllCollectionInfoToCompactRequest,
+    ) -> Result<Response<GetAllCollectionInfoToCompactResponse>, Status> {
+        // TODO(rescrv):  Realistically we could make this configurable.
+        const MAX_COLLECTION_INFO_NUMBER: usize = 10000;
+        let mut selected_rollups = Vec::with_capacity(MAX_COLLECTION_INFO_NUMBER);
+        // Do a non-allocating pass here.
+        {
+            let need_to_compact = self.need_to_compact.lock();
+            for (collection_id, rollup) in need_to_compact.iter() {
+                if rollup.limit_log_position >= rollup.start_log_position
+                    && rollup.limit_log_position - rollup.start_log_position
+                        >= request.min_compaction_size
+                {
+                    selected_rollups.push((*collection_id, *rollup));
+                }
+            }
+        }
+        // Then allocate the collection ID strings outside the lock.
+        let mut all_collection_info = Vec::with_capacity(selected_rollups.len());
+        for (collection_id, rollup) in selected_rollups {
+            all_collection_info.push(CollectionInfo {
+                collection_id: collection_id.to_string(),
+                first_log_offset: rollup.start_log_position.offset() as i64,
+                first_log_ts: rollup.start_log_position.offset() as i64,
+            });
+        }
+        Ok(Response::new(GetAllCollectionInfoToCompactResponse {
+            all_collection_info,
+        }))
+    }
+
+    /// Read a prefix of the dirty log, coalescing records as it goes.
+    ///
+    /// This will rewrite the dirty log's coalesced contents at the tail and adjust the cursor to
+    /// said position so that the next read is O(1) if there are no more writes.
+    #[tracing::instrument(skip(self), err(Display))]
+    async fn roll_dirty_log(&self) -> Result<(), Error> {
+        // Ensure at most one request at a time.
+        let _guard = self.rolling_up.lock().await;
+        let (witness, cursor, dirty_markers) = self.read_dirty_log().await?;
+        self.metrics
```

Comment:
should this metric be fired after the rewrite? Also its not the number rewritten right?